### PR TITLE
Allows Mongo client settings customization

### PR DESCRIPTION
**Pull request by cescoffier**
Also, document how this feature can be used to configure CSFLE.

### CODE REVIEW

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java [hunk 1]
```diff
@@ -433,4 +439,38 @@ void runtimeInitializedClasses(BuildProducer<RuntimeInitializedClassBuildItem> r
...
+    /**
+     * Ensure we have at most one customizer per Mongo client.
+     *
+     * @param beans the beans
+     * @param validation the producer used to report issues
+     */
+    @BuildStep
+    void validateMongoConfigCustomizers(BeanDiscoveryFinishedBuildItem beans,
+            BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> validation) {
+        var type = DotName.createSimple(MongoClientCustomizer.class);
```
Comment (loicmathieu) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:452]: We usually declare DotName as a constant, can you move it near to the other as constant at the top of the class?

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java [hunk 2]
```diff
@@ -433,4 +439,38 @@ void runtimeInitializedClasses(BuildProducer<RuntimeInitializedClassBuildItem> r
...
+        HashMap<String, List<String>> customizers = new HashMap<>();
+
+        for (BeanInfo bean : beans.getBeans()) {
+            if (bean.hasType(type)) {
+                var name = bean.getQualifier(DotName.createSimple(MongoClientName.class));
```
Comment (loicmathieu) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:457]: This DotName already exist as a constant

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java [hunk 3]
```diff
@@ -326,9 +328,41 @@ private MongoClientSettings createMongoConfiguration(MongoClientConfig config) {
...
+            if (qualifier.annotationType().equals(MongoClientName.class)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private MongoClientSettings.Builder customize(String name, MongoClientSettings.Builder settings) {
+        // If there is no customizer for the default client, use the first one that does not have a client name qualifier
+        // Otherwise look for the right customizer
```
Comment (loicmathieu) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java:347]: This comment seems not accurate, we use the first one that does not have a client name qualifier, if not we don't apply any customizer. The second line seems to be misleading.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java [hunk 4]
```diff
@@ -326,9 +328,41 @@ private MongoClientSettings createMongoConfiguration(MongoClientConfig config) {
...
+        // At build time, we check that we have at most one customizer per client.
+        if (MongoClientBeanUtil.isDefault(name)) {
+            var list = customizers.handlesStream().filter(h -> doesNotHaveClientNameQualifier(h.getBean()))
+                    .collect(Collectors.toList());
```
Comment (loicmathieu) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java:351]: You can use findFirst() here to slightly improve readability but it's not mandatory, you can leave it as is.

**Review by loicmathieu — APPROVED** [review]
Hi,
A few small remarks but overall LGTM

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java (hunk 3 above)
Comment (cescoffier) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java]: One day, I will explain you that comment... It is going to crush your deepest dreams. 

I fixed it.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java (hunk 4 above)
Comment (cescoffier) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java]: I switched to findFirst.

It's what I was using initially, but I was not happy with the maybe.get().get().customize(....). The first `get` retrieve the value from the optional. The second `get` retrieves the bean instance.

Comment (loicmathieu) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java]: Well, I said 'can' so you could have keep it as is ;)

**Pull request merged**